src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java

```diff
@@ -0,0 +1,203 @@
+    consumeService.awaitShutdown();
+  }
+
+  @Test
+  public void lifecycleTest() throws ExecutionException, InterruptedException {
+    ConsumeService consumeService = consumeService();
+
+    /* Nothing should be started */
+    Assert.assertFalse(consumeService.isRunning());
+    Assert.assertNotNull(consumeService.getServiceName());
+
+    /* Should accept but ignore start because start has not been called */
+    consumeService.stop();
+    Assert.assertFalse(consumeService.isRunning());
+
+    /* Should start */
+    consumeService.start();
+    Assert.assertTrue(consumeService.isRunning());
+
+    /* Should allow start to be called more than once */
+    consumeService.stop();
+    consumeService.stop();
+    Assert.assertFalse(consumeService.isRunning());
+
+    /* Should be allowed to shutdown more than once. */
+    consumeService.awaitShutdown();
+    consumeService.awaitShutdown();
+    Assert.assertFalse(consumeService.isRunning());
+  }
+
+
+  private ConsumeService consumeService() throws ExecutionException, InterruptedException {
+    /* Sample ConsumeService instance for unit testing */
+
+    LOG.info("Creating an instance of Consume Service for testing..");
+    Map<String, Object> fakeProps = new HashMap<>();
+
+    fakeProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BROKER_LIST);
+    fakeProps.put(CommonServiceConfig.ZOOKEEPER_CONNECT_CONFIG, ZK_CONNECT);
+    fakeProps.put(CommonServiceConfig.TOPIC_CONFIG, TOPIC);
+
+    ConsumerFactory consumerFactory = new FakeConsumerFactory(fakeProps);
+    CompletableFuture<Void> topicPartitionResult = new CompletableFuture<>();
+    topicPartitionResult.complete(null);
+
+    return new ConsumeService("name", topicPartitionResult, consumerFactory);
+  }
+
+  @Test
+  public void awaitShutdownOtherThread() throws Exception {
+    final ConsumeService consumeService = consumeService();
+    final AtomicReference<Throwable> error = new AtomicReference<>();
+
+    Thread thread = new Thread("test awaitshutdown thread") {
+      @Override
+      public void run() {
+        try {
+          consumeService.awaitShutdown();
+        } catch (Throwable t) {
+          error.set(t);
+        }
+      }
+    };
+
+    thread.start();
+    consumeService.start();
+    Thread.sleep(100);
+
+    consumeService.stop();
+    thread.join(500);
+
+    Assert.assertFalse(thread.isAlive());
+    Assert.assertEquals(error.get(), null);
+
+  }
+
+  static final class FakeConsumerFactory implements ConsumerFactory {
+    final Map<String, Object> _props;
+
+    /* Required */
+    public FakeConsumerFactory(Map<String, Object> props) {
+      _props = props;
+    }
+
+    @Override
+    public AdminClient adminClient() {
+      return AdminClient.create(_props);
+    }
+
+    @Override
+    public int latencySlaMs() {
+      return 20000;
+    }
+
+    @Override
+    public KMBaseConsumer baseConsumer() {
+      return Mockito.mock(KMBaseConsumer.class);
```

Comment:
How are expectations being added to this mocked consumer?  Shouldn't it return some messages or something.
This might be a little dated but it seems fairly comprehensive.
https://www.vogella.com/tutorials/Mockito/article.html

Reply:
_Updated._